betterproto/tests/test_get_ref_type.py

```diff
@@ -77,6 +79,227 @@ def test_importing_google_wrappers_unwraps_them(google_type: str, expected_name:
+    imports = set()
+    name = get_ref_type(
+        package="", imports=imports, source_type="child_package.example_message"
+    )
+
+    assert imports == {"from . import child_package"}
+    assert name == "child_package.ExampleMessage"
+
+
+def test_import_nested_child_from_root():
+    imports = set()
+    name = get_ref_type(package="", imports=imports, source_type="nested.child.Message")
+
+    assert imports == {"from .nested import child as nested_child"}
+    assert name == "nested_child.Message"
+
+
+def test_import_deeply_nested_child_from_root():
+    imports = set()
+    name = get_ref_type(
+        package="", imports=imports, source_type="deeply.nested.child.Message"
+    )
+
+    assert imports == {"from .deeply.nested import child as deeply_nested_child"}
+    assert name == "deeply_nested_child.Message"
+
+
+def test_import_deeply_nested_child_from_package():
+    imports = set()
+    name = get_ref_type(
+        package="package",
+        imports=imports,
+        source_type="package.deeply.nested.child.Message",
+    )
+
+    assert imports == {"from .deeply.nested import child as deeply_nested_child"}
+    assert name == "deeply_nested_child.Message"
+
+
+def test_import_root_sibling():
+    imports = set()
+    name = get_ref_type(package="", imports=imports, source_type="Message")
+
+    assert imports == set()
+    assert name == '"Message"'
+
+
+def test_import_nested_siblings():
+    imports = set()
+    name = get_ref_type(package="foo", imports=imports, source_type="foo.Message")
+
+    assert name == '"Message"'
+
+
+def test_import_deeply_nested_siblings():
+    imports = set()
+    name = get_ref_type(
+        package="foo.bar", imports=imports, source_type="foo.bar.Message"
+    )
+
+    assert name == '"Message"'
+
+
+def test_import_parent_package_from_child():
+    imports = set()
+    name = get_ref_type(
+        package="package.child", imports=imports, source_type="package.Message"
+    )
+
+    assert imports == {"from ... import package as __package__"}
+    assert name == "__package__.Message"
+
+
+def test_import_parent_package_from_deeply_nested_child():
+    imports = set()
+    name = get_ref_type(
+        package="package.deeply.nested.child",
+        imports=imports,
+        source_type="package.deeply.nested.Message",
+    )
+
+    assert imports == {"from ... import nested as __nested__"}
+    assert name == "__nested__.Message"
+
+
+def test_import_ancestor_package_from_nested_child():
+    imports = set()
+    name = get_ref_type(
+        package="package.ancestor.nested.child",
+        imports=imports,
+        source_type="package.ancestor.Message",
+    )
+
+    assert imports == {"from .... import ancestor as ___ancestor__"}
+    assert name == "___ancestor__.Message"
+
+
+def test_import_root_package_from_child():
+    imports = set()
+    name = get_ref_type(package="package.child", imports=imports, source_type="Message")
+
+    assert imports == {"from ... import Message"}
```

Comment:
This case looks like it follows a different rule to the previous test case with respect to importing the renamed module? Is this avoidable?

Reply:
Somewhat yes and no.

What I tried to do wherever possible was alias the modules, but not the imported types.
Although aliases look ugly, at least the type names are preserved this way.

For ancestor imports, there is no identified package to import that wraps the Type, so you cannot give it an alias. e.g, this is not accepted by python:

```python
from ... import . as ___root__

___root__.Message
```

1. If you know for sure that the ancestor is not the root of the package, you can import 1 level higher up, and alias that. This is what i have done in `reference_ancestor` https://github.com/danielgtaylor/python-betterproto/blob/e2d672a422682cb631bfe772f5c51d0d98b02126/betterproto/compile/importing.py#L126-L135  
  - However this approach is unsafe when the target type is in the top level package, as its not guaranteed that the generated code resides in its own package, so we might end up ascending outside of the package
  - So for the case you mentioned, this approach does not work safely.
2. You could alias the type itself. e.g.
   `from ... import Message as ___Message__`, or  
   `from ... import Message as ___root_Message_`

2 would be a safe and valid solution that prevents name collisions when importing 'Message' from root into a subpackage that also contains 'Message'.

Unfortunately, it makes the name ugly, in that case.

In the future, I hope to improve the importing logic such that it keeps track of occupied names per package, and only creates aliases when necessary, although this will complicate things if in the future we want to support parallel compilation (e.g. multiple protoc processes compiling a single package hierarchy).

Reply:
I think robustness should be prioritized over the prettiness of the names in general. The first goal of betterproto should be that it works. Then, that it generates beautiful code. So if you agree, I will also alias the root types.

Reply:
>    `from ... import Message as ___root_Message_`

Until someone calls a package root 🙃

Reply:
Good point 😄 

I've aliased them now as `___Message__`, I don't think adding `root` will help a lot in the end